### PR TITLE
Implement ESDT functions

### PR DESCRIPTION
**Pull request by bbyalcinkaya**

### CODE REVIEW

File: elrond-config.md [hunk 1]
```diff
@@ -538,37 +551,160 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
...
       [priority(60)]
 
-    rule <commands> callContract(FROM, TO, VALUE, FUNCNAME:WasmStringToken, ARGS, GASLIMIT, GASPRICE)
+    // The memory instance linked to the contract's account should be restored to the initial state after the contract call.
+    rule <commands> callContract(FROM, TO, VALUE, ESDT, FUNCNAME:WasmStringToken, ARGS, GASLIMIT, GASPRICE)
                  => pushWorldState
                  ~> transferFunds(FROM, TO, VALUE)
-                 ~> mkCall(FROM, TO, VALUE, FUNCNAME, ARGS, GASLIMIT, GASPRICE)
+                 ~> transferESDTs(FROM, TO, ESDT)
+                 ~> mkCall(FROM, TO, VALUE, ESDT, FUNCNAME, ARGS, GASLIMIT, GASPRICE)
```
Comment (yzhang90) [inline elrond-config.md:647]: Does the memory need to be emptied when make a call?

Comment (bbyalcinkaya) [reply elrond-config.md:647]: @yzhang90 
When a module is initialized (during SC deployment), the memory consists of constants like static strings and empty sections. This rule assumes the memory is in this initial state, so it does not empty it _before_ the `mkCall`.

We only clean up the memory after `#endWasm` is reached using the `#restoreMem(_)` helper.

File: elrond-config.md [hunk 2]
```diff
@@ -538,37 +551,165 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
...
+      [priority(60)]
+
+    rule <commands> transferESDT(ACCTFROM, ACCTTO, esdtTransfer(TOKEN, VALUE, 0)) 
+                 => . ... 
+         </commands>
+         <account>
+           <address> ACCTFROM </address>
+           <esdtData>
+             <esdtId> TOKEN </esdtId>
+             <esdtBalance> ORIGFROM => ORIGFROM -Int VALUE </esdtBalance>
```
[the remark this answers is not in the document]
Comment (bbyalcinkaya) [reply elrond-config.md:589]: Go implementation leaves it.

File: vmhooks/eei-helpers.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  rule #idMinLen        => 10
+  rule #idMaxLen        => 17
+
+  syntax Bool ::= #validateToken( Bytes )                [function, total]
+ // -----------------------------------------------------------------
+  rule #validateToken(Bs) => false requires lengthBytes(Bs) <Int #idMinLen
+                                     orBool lengthBytes(Bs) >Int #idMaxLen
+  rule #validateToken(Bs) => #isTickerValid( #getTicker(Bs) ) 
+                     andBool #randomCharsAreValid( #getRandomChars(Bs) )
+                     andBool Bs[(lengthBytes(Bs) -Int #randomCharsLen) -Int 1 ] ==Int ordChar("-")
```
[the remark this answers is not in the document]
Comment (bbyalcinkaya) [reply vmhooks/eei-helpers.md:31]: @virgil-serbanuta `#idMinLen` and `#randomCharsLen` are macros to constant ints. I changed the rule for `#idMinLen` as follows. Is it still possible to become undefined?

```
rule #idMinLen  => #tickerMinLen +Int #randomCharsLen +Int 1
```

([Go code](https://github.com/runtimeverification/elrond-semantics/pull/95#discussion_r1174971529))

File: elrond-config.md (hunk 1 above)
Comment (bbyalcinkaya) [reply elrond-config.md:656]: > How do calls between contracts work? I'm assuming that the called contract needs its own "fresh" memory.

Added a TODO comment about this. Each contract call should run its own wasm module instance, and there should be a `<callStack>` to keep intermediate `<callState>`'s.

**Review by virgil-serbanuta — COMMENTED** [review]
Looks good to me. Please check the latest comments.

I think that Yi had some unsolved comments, so I will let him approve.

File: vmhooks/eei-helpers.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                              requires lengthBytes(Bs) >=Int #idMinLen
+                               andBool lengthBytes(Bs) <=Int #idMaxLen
+
+  syntax Bytes ::= #getTicker(Bytes)                    [function, total]
+                 | #getRandomChars(Bytes)               [function, total]
+ // ------------------------------------------------------------------------------------
+  rule #getTicker(Bs) => substrBytes(Bs, 0, lengthBytes(Bs) -Int #randomCharsLen -Int 1) 
+    requires lengthBytes(Bs) >=Int #randomCharsLen
```
[the remark this answers is not in the document]
Comment (bbyalcinkaya) [reply vmhooks/eei-helpers.md]: done!

File: elrond-config.md [hunk 5]
```diff
@@ -538,37 +551,160 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
       requires ACCTFROM =/=K ACCTTO andBool VALUE <=Int ORIGFROM
       [priority(60)]
 
+    // transferESDTs performs multiple ESDT transfers and finally returns #transferSuccess
```
[the remark this answers is not in the document]
Comment (bbyalcinkaya) [reply elrond-config.md]: Added TODO comments for the failure cases above the `rule <commands> transferESDT(...) => . ... </commands>` rule

**Review by yzhang90 — APPROVED** [review]
LGTM

**Pull request merged**